src/transformStateWithClones.js

```diff
@@ -7,7 +7,46 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const history = [];
+  let previousState = state;
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        previousState = combineProperties(previousState, action.extraData);
+        break;
+      case 'removeProperties':
+        previousState = removeProperties(previousState, action.keysToRemove);
+        break;
+      case 'clear':
+        previousState = clearState(previousState);
+        break;
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+
+    history.push(previousState);
```

Comment:
You're pushing the modified state into the history array after each action. However, as per the checklist, you should make a copy of the state after the switch statement and before pushing it into the history array to ensure you're not repeating yourself.